File: DFS/Lint_614.java
// https://www.lintcode.com/problem/binary-tree-longest-consecutive-sequence-ii/description?_from=ladder&&fromId=1

// Thinking:
// record [increase, decrease] value for each node,
// which refer to the maximum inc or dec value from current node to below nodes
// by comparing whether the current node value is consecutive with its child node,
// update the child.inc + 1 or child.dec + 1 and pick the max inc and dec
// then the max length from current node = inc + dec - 1;

// M1, ResultType
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    class ResultType {
        int maxLen, inr, dcr;
        public ResultType(int maxLen, int inr, int dcr) {
            this.maxLen = maxLen;
            this.inr = inr;
            this.dcr = dcr;
        }
    }
    public int longestConsecutive(TreeNode root) {
        return helper(root).maxLen;
    }
    
    private ResultType helper(TreeNode root) {
        if (root == null)
            return new ResultType(0, 0, 0);
        
        ResultType left = helper(root.left);
        ResultType right = helper(root.right);
        
        int inr = 1, dcr = 1; 
        if (root.left != null) {
            if (root.left.val + 1 == root.val) 
                inr = left.inr + 1;
            
            if (root.left.val - 1 == root.val)
                dcr = left.dcr + 1;
        }
        
        if (root.right != null) {
            if (root.right.val + 1 == root.val)
                inr = Math.max(inr, right.inr + 1);
            
            if (root.right.val - 1 == root.val)
                dcr = Math.max(dcr, right.dcr + 1);
        }
        
        int len = inr + dcr - 1;
        len = Math.max(len, Math.max(left.maxLen, right.maxLen));
        
        return new ResultType(len, inr, dcr);
    }
}

File: 5 DFS_py/Lint_614.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: the root of binary tree
    @return: the length of the longest consecutive sequence path
    """
    def longestConsecutive2(self, root):
        up, down, max_len = self.helper(root)
        return max_len
        
    def helper(self, root):
        if root is None:
            return 0, 0, 0
            
        l_up, l_down, l_max_len = self.helper(root.left)
        r_up, r_down, r_max_len = self.helper(root.right)
        
        down, up = 0, 0
        
        if root.left:
            if root.val + 1 == root.left.val:
                down = max(down, l_down + 1)
            if root.val - 1 == root.left.val:
                up = max(up, l_up + 1)
                
        if root.right:
            if root.val + 1 == root.right.val:
                down = max(down, r_down + 1)
            if root.val - 1 == root.right.val:
                up = max(up, r_up + 1)
                
        max_len = max(1 + down + up, l_max_len, r_max_len)
        
        return up, down, max_len
        
            
        
